src/WebJobs.Extensions.SendGrid/Config/SendGridConfiguration.cs

```diff
@@ -68,28 +66,35 @@ public void Initialize(ExtensionConfigContext context)
                 throw new ArgumentNullException("context");
             }
 
-            INameResolver nameResolver = context.Config.NameResolver;
-            _defaultApiKey = nameResolver.Resolve(AzureWebJobsSendGridApiKeyName);
+            var metadata = new ConfigMetadata();
+            context.ApplyConfig(metadata, "sendgrid");
+            this.ToAddress = SendGridHelpers.Apply(this.ToAddress, metadata.To);
+            this.FromAddress = SendGridHelpers.Apply(this.FromAddress, metadata.From);
 
-            IConverterManager converterManager = context.Config.GetService<IConverterManager>();
-            converterManager.AddConverter<string, Mail>(SendGridHelpers.CreateMessage);
-            converterManager.AddConverter<JObject, Mail>(SendGridHelpers.CreateMessage);
-
-            BindingFactory factory = new BindingFactory(nameResolver, converterManager);
-            IBindingProvider outputProvider = factory.BindToCollector<SendGridAttribute, Mail>((attr) =>
+            if (string.IsNullOrEmpty(this.ApiKey))
             {
-                string apiKey = FirstOrDefault(attr.ApiKey, ApiKey, _defaultApiKey);
-                ISendGridClient sendGrid = _sendGridClientCache.GetOrAdd(apiKey, a => ClientFactory.Create(a));
-                return new SendGridMailAsyncCollector(this, attr, sendGrid);
-            });
+                INameResolver nameResolver = context.Config.NameResolver;
+                this.ApiKey = nameResolver.Resolve(AzureWebJobsSendGridApiKeyName);
+            }
 
-            IExtensionRegistry extensions = context.Config.GetService<IExtensionRegistry>();
-            extensions.RegisterBindingRules<SendGridAttribute>(ValidateBinding, nameResolver, outputProvider);
+            context
+                .AddConverter<string, Mail>(SendGridHelpers.CreateMessage)
+                .AddConverter<JObject, Mail>(SendGridHelpers.CreateMessage)
+                .AddBindingRule<SendGridAttribute>()
+                    .BindToCollector<Mail>(CreateCollector);
+        }
+
+        private IAsyncCollector<Mail> CreateCollector(SendGridAttribute attr)
```

Comment:
Per invocation. That's because the attribute can have { } which are resolved to invocation specific items.